ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,654 @@
+ *
+ *      Redistributions in binary form must reproduce the above
+ *      copyright notice, this list of conditions and the following
+ *      disclaimer in the documentation and/or other materials provided
+ *      with the distribution.
+ *
+ *      Neither the name of Sandia nor the names of any contributors may
+ *      be used to endorse or promote products derived from this software
+ *      without specific prior written permission.
+ *
+ *      Neither the name of Open Grid Computing nor the names of any
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ *      Modified source versions must be plainly marked as such, and
+ *      must not be misrepresented as being the original software.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdint.h>
+#include <stdbool.h>
+#include <limits.h>
+#include <string.h>
+#include <dirent.h>
+#include <coll/rbt.h>
+#include <coll/idx.h>
+#include <sys/queue.h>
+#include <time.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "ldms_jobid.h"
+
+#define _GNU_SOURCE
+
+#define SAMP "lustre_client"
+#define LLITE_PATH "/proc/fs/lustre/llite"
+#define DEFAULT_SCHEMA_NAME "lustre_client"
+/* from lustre code */
+#define MAX_OBD_NAME 128
+#define LUSTRE_MAXFSNAME 8
+
+static ldmsd_msg_log_f log_fn;
+static char producer_name[LDMS_PRODUCER_NAME_MAX+1];
+static char schema_name[LDMS_SET_NAME_MAX+1];
+static char host_name[HOST_NAME_MAX+1];
+static uint64_t component_id;
+static time_t last_refresh;
+static time_t refresh_interval = 30;
+static int job_id_index;
+LJI_GLOBALS;
+
+/* red-black tree root for llites */
+static struct rbt llite_tree;
+
+struct llite_data {
+        char *fs_name;
+        char *name;
+        char *path;
+        char *stats_path;
+        ldms_set_t general_metric_set; /* a pointer */
+        struct rbn llite_tree_node;
+};
+
+static ldms_schema_t llite_general_schema;
+
+static char *llite_stats_uint64_t_entries[] = {
+        "dirty_pages_hits",
+        "dirty_pages_misses",
+        "read_bytes.sum",
+        "write_bytes.sum",
+        "brw_read.sum",
+        "brw_write.sum",
```

Comment:
Does the .sum mean that the value is cumulative? Are all of these values sums, or are the other values rates?